File: .spikes/rem-db/src/rem_db/predicates.py
"""SQL-like predicates for filtering."""

from dataclasses import dataclass
from typing import Any

import numpy as np

from .models import Entity, Order, Resource


@dataclass
class Predicate:
    """Base predicate class."""

    def evaluate(self, obj: Entity | Resource) -> bool:
        """Evaluate predicate against object."""
        raise NotImplementedError


@dataclass
class Eq(Predicate):
    """field == value"""

    field: str
    value: Any

    def evaluate(self, obj: Entity | Resource) -> bool:
        val = self._get_field(obj, self.field)
        return val == self.value

    @staticmethod
    def _get_field(obj: Entity | Resource, field: str) -> Any:
        # Check direct attributes
        if hasattr(obj, field):
            return getattr(obj, field)
        # Check properties/metadata
        if hasattr(obj, "properties") and field in obj.properties:
            return obj.properties[field]
        if hasattr(obj, "metadata") and field in obj.metadata:
            return obj.metadata[field]
        return None


@dataclass
class Ne(Predicate):
    """field != value"""

    field: str
    value: Any

    def evaluate(self, obj: Entity | Resource) -> bool:
        val = Eq._get_field(obj, self.field)
        return val != self.value


@dataclass
class Gt(Predicate):
    """field > value"""

    field: str
    value: Any

    def evaluate(self, obj: Entity | Resource) -> bool:
        val = Eq._get_field(obj, self.field)
        return val is not None and val > self.value


@dataclass
class Gte(Predicate):
    """field >= value"""

    field: str
    value: Any

    def evaluate(self, obj: Entity | Resource) -> bool:
        val = Eq._get_field(obj, self.field)
        return val is not None and val >= self.value


@dataclass
class Lt(Predicate):
    """field < value"""

    field: str
    value: Any

    def evaluate(self, obj: Entity | Resource) -> bool:
        val = Eq._get_field(obj, self.field)
        return val is not None and val < self.value


@dataclass
class Lte(Predicate):
    """field <= value"""

    field: str
    value: Any

    def evaluate(self, obj: Entity | Resource) -> bool:
        val = Eq._get_field(obj, self.field)
        return val is not None and val <= self.value


@dataclass
class In(Predicate):
    """field IN [values]"""

    field: str
    values: list[Any]

    def evaluate(self, obj: Entity | Resource) -> bool:
        val = Eq._get_field(obj, self.field)
        return val in self.values


@dataclass
class NotIn(Predicate):
    """field NOT IN [values]"""

    field: str
    values: list[Any]

    def evaluate(self, obj: Entity | Resource) -> bool:
        val = Eq._get_field(obj, self.field)
        return val not in self.values


@dataclass
class Contains(Predicate):
    """field CONTAINS substring"""

    field: str
    substring: str

    def evaluate(self, obj: Entity | Resource) -> bool:
        val = Eq._get_field(obj, self.field)
        return isinstance(val, str) and self.substring in val


@dataclass
class StartsWith(Predicate):
    """field STARTS WITH prefix"""

    field: str
    prefix: str

    def evaluate(self, obj: Entity | Resource) -> bool:
        val = Eq._get_field(obj, self.field)
        return isinstance(val, str) and val.startswith(self.prefix)


@dataclass
class EndsWith(Predicate):
    """field ENDS WITH suffix"""

    field: str
    suffix: str

    def evaluate(self, obj: Entity | Resource) -> bool:
        val = Eq._get_field(obj, self.field)
        return isinstance(val, str) and val.endswith(self.suffix)


@dataclass
class And(Predicate):
    """pred1 AND pred2 AND ..."""

    predicates: list[Predicate]

    def evaluate(self, obj: Entity | Resource) -> bool:
        return all(p.evaluate(obj) for p in self.predicates)


@dataclass
class Or(Predicate):
    """pred1 OR pred2 OR ..."""

    predicates: list[Predicate]

    def evaluate(self, obj: Entity | Resource) -> bool:
        return any(p.evaluate(obj) for p in self.predicates)


@dataclass
class Not(Predicate):
    """NOT pred"""

    predicate: Predicate

    def evaluate(self, obj: Entity | Resource) -> bool:
        return not self.predicate.evaluate(obj)


@dataclass
class Exists(Predicate):
    """field IS NOT NULL"""

    field: str

    def evaluate(self, obj: Entity | Resource) -> bool:
        return Eq._get_field(obj, self.field) is not None


@dataclass
class NotExists(Predicate):
    """field IS NULL"""

    field: str

    def evaluate(self, obj: Entity | Resource) -> bool:
        return Eq._get_field(obj, self.field) is None


@dataclass
class All(Predicate):
    """Always true"""

    def evaluate(self, obj: Entity | Resource) -> bool:
        return True


@dataclass
class Query:
    """Query builder with predicates."""

    predicate: Predicate = None
    order_by: tuple[str, Order] | None = None
    limit: int | None = None
    offset: int | None = None

    def __post_init__(self):
        if self.predicate is None:
            self.predicate = All()

    def filter(self, predicate: Predicate) -> "Query":
        """Add filter predicate."""
        if isinstance(self.predicate, All):
            self.predicate = predicate
        else:
            self.predicate = And([self.predicate, predicate])
        return self

    def sort(self, field: str, order: Order = Order.ASC) -> "Query":
        """Add sorting."""
        self.order_by = (field, order)
        return self

    def take(self, n: int) -> "Query":
        """Limit results."""
        self.limit = n
        return self

    def skip(self, n: int) -> "Query":
        """Skip results."""
        self.offset = n
        return self

File: percolate-rust/src/memory/predicates.rs
//! Predicate-based filtering for REM queries
//!
//! Provides SQL-like predicates that can be combined and evaluated against entities.

use serde::{Deserialize, Serialize};
use serde_json::Value;

/// Predicate for filtering entities and resources
#[derive(Debug, Clone, Serialize, Deserialize)]
pub enum Predicate {
    // Comparison operations
    /// field == value
    Eq(String, Value),
    /// field != value
    Ne(String, Value),
    /// field > value
    Gt(String, Value),
    /// field >= value
    Gte(String, Value),
    /// field < value
    Lt(String, Value),
    /// field <= value
    Lte(String, Value),

    // Set operations
    /// field IN [values]
    In(String, Vec<Value>),
    /// field NOT IN [values]
    NotIn(String, Vec<Value>),

    // String operations
    /// field CONTAINS substring
    Contains(String, String),
    /// field STARTS WITH prefix
    StartsWith(String, String),
    /// field ENDS WITH suffix
    EndsWith(String, String),
    /// field MATCHES regex
    Matches(String, String),

    // Logical operations
    /// pred1 AND pred2 AND ...
    And(Vec<Predicate>),
    /// pred1 OR pred2 OR ...
    Or(Vec<Predicate>),
    /// NOT pred
    Not(Box<Predicate>),

    // Vector operations
    /// Semantic similarity search
    VectorSimilar {
        field: String,
        query: Vec<f32>,
        top_k: usize,
        min_score: f32,
    },

    // Existence checks
    /// field IS NOT NULL
    Exists(String),
    /// field IS NULL
    NotExists(String),

    // Always true/false (for composition)
    All,
    None,
}

impl Predicate {
    /// Evaluate predicate against a value map
    ///
    /// # Example
    ///
    /// ```rust
    /// use percolate_rust::memory::predicates::Predicate;
    /// use serde_json::{json, Value};
    /// use std::collections::HashMap;
    ///
    /// let mut fields = HashMap::new();
    /// fields.insert("status".to_string(), json!("active"));
    /// fields.insert("age".to_string(), json!(25));
    ///
    /// let pred = Predicate::Eq("status".into(), json!("active"));
    /// assert!(pred.evaluate(&fields));
    /// ```
    pub fn evaluate(&self, fields: &std::collections::HashMap<String, Value>) -> bool {
        match self {
            Predicate::Eq(field, value) => {
                fields.get(field).map(|v| v == value).unwrap_or(false)
            }
            Predicate::Ne(field, value) => {
                fields.get(field).map(|v| v != value).unwrap_or(true)
            }
            Predicate::Gt(field, value) => {
                fields.get(field).map(|v| compare_values(v, value) == std::cmp::Ordering::Greater).unwrap_or(false)
            }
            Predicate::Gte(field, value) => {
                fields.get(field).map(|v| {
                    let cmp = compare_values(v, value);
                    cmp == std::cmp::Ordering::Greater || cmp == std::cmp::Ordering::Equal
                }).unwrap_or(false)
            }
            Predicate::Lt(field, value) => {
                fields.get(field).map(|v| compare_values(v, value) == std::cmp::Ordering::Less).unwrap_or(false)
            }
            Predicate::Lte(field, value) => {
                fields.get(field).map(|v| {
                    let cmp = compare_values(v, value);
                    cmp == std::cmp::Ordering::Less || cmp == std::cmp::Ordering::Equal
                }).unwrap_or(false)
            }
            Predicate::In(field, values) => {
                fields.get(field).map(|v| values.contains(v)).unwrap_or(false)
            }
            Predicate::NotIn(field, values) => {
                fields.get(field).map(|v| !values.contains(v)).unwrap_or(true)
            }
            Predicate::Contains(field, substring) => {
                fields.get(field).and_then(|v| v.as_str()).map(|s| s.contains(substring)).unwrap_or(false)
            }
            Predicate::StartsWith(field, prefix) => {
                fields.get(field).and_then(|v| v.as_str()).map(|s| s.starts_with(prefix)).unwrap_or(false)
            }
            Predicate::EndsWith(field, suffix) => {
                fields.get(field).and_then(|v| v.as_str()).map(|s| s.ends_with(suffix)).unwrap_or(false)
            }
            Predicate::Matches(field, pattern) => {
                fields.get(field).and_then(|v| v.as_str()).map(|s| {
                    regex::Regex::new(pattern)
                        .map(|re| re.is_match(s))
                        .unwrap_or(false)
                }).unwrap_or(false)
            }
            Predicate::And(predicates) => {
                predicates.iter().all(|p| p.evaluate(fields))
            }
            Predicate::Or(predicates) => {
                predicates.iter().any(|p| p.evaluate(fields))
            }
            Predicate::Not(predicate) => {
                !predicate.evaluate(fields)
            }
            Predicate::VectorSimilar { .. } => {
                // Vector similarity requires HNSW index lookup, not evaluated here
                // This is handled at the query execution level
                true
            }
            Predicate::Exists(field) => {
                fields.contains_key(field)
            }
            Predicate::NotExists(field) => {
                !fields.contains_key(field)
            }
            Predicate::All => true,
            Predicate::None => false,
        }
    }
}

/// Compare two JSON values for ordering
fn compare_values(a: &Value, b: &Value) -> std::cmp::Ordering {
    use std::cmp::Ordering;

    match (a, b) {
        (Value::Number(a), Value::Number(b)) => {
            if let (Some(a_f), Some(b_f)) = (a.as_f64(), b.as_f64()) {
                a_f.partial_cmp(&b_f).unwrap_or(Ordering::Equal)
            } else {
                Ordering::Equal
            }
        }
        (Value::String(a), Value::String(b)) => a.cmp(b),
        (Value::Bool(a), Value::Bool(b)) => a.cmp(b),
        _ => Ordering::Equal,
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;
    use std::collections::HashMap;

    #[test]
    fn test_eq_predicate() {
        let mut fields = HashMap::new();
        fields.insert("status".to_string(), json!("active"));

        let pred = Predicate::Eq("status".to_string(), json!("active"));
        assert!(pred.evaluate(&fields));

        let pred = Predicate::Eq("status".to_string(), json!("inactive"));
        assert!(!pred.evaluate(&fields));
    }

    #[test]
    fn test_gt_predicate() {
        let mut fields = HashMap::new();
        fields.insert("age".to_string(), json!(25));

        let pred = Predicate::Gt("age".to_string(), json!(18));
        assert!(pred.evaluate(&fields));

        let pred = Predicate::Gt("age".to_string(), json!(30));
        assert!(!pred.evaluate(&fields));
    }

    #[test]
    fn test_and_predicate() {
        let mut fields = HashMap::new();
        fields.insert("status".to_string(), json!("active"));
        fields.insert("age".to_string(), json!(25));

        let pred = Predicate::And(vec![
            Predicate::Eq("status".to_string(), json!("active")),
            Predicate::Gt("age".to_string(), json!(18)),
        ]);
        assert!(pred.evaluate(&fields));
    }

    #[test]
    fn test_in_predicate() {
        let mut fields = HashMap::new();
        fields.insert("status".to_string(), json!("active"));

        let pred = Predicate::In(
            "status".to_string(),
            vec![json!("active"), json!("pending")],
        );
        assert!(pred.evaluate(&fields));
    }
}
